drivers/sensor/nuvoton_adc_cmp_npcx/adc_cmp_npcx.c

```diff
@@ -0,0 +1,274 @@
+				      bool is_mv, uint32_t value)
+{
+	const struct adc_cmp_npcx_config *const config = dev->config;
+	struct adc_npcx_threshold_param param;
+	int ret;
+
+	param.type = ADC_NPCX_THRESHOLD_PARAM_THVAL;
+	if (is_mv) {
+		ret = adc_npcx_threshold_mv_to_thrval(value, &param.val);
+		if (ret) {
+			return ret;
+		}
+	}
+
+	ret = adc_npcx_threshold_ctrl_set_param(config->adc,
+						config->th_sel, &param);
+	if (ret) {
+		return ret;
+	}
+
+	param.type = ADC_NPCX_THRESHOLD_PARAM_L_H;
+	param.val = is_upper ? ADC_NPCX_THRESHOLD_PARAM_L_H_HIGHER :
+		ADC_NPCX_THRESHOLD_PARAM_L_H_LOWER;
+
+	ret = adc_npcx_threshold_ctrl_set_param(config->adc,
+						config->th_sel, &param);
+
+	return ret;
+}
+
+static int adc_cmp_npcx_attr_set(const struct device *dev,
+			  enum sensor_channel chan,
+			  enum sensor_attribute attr,
+			  const struct sensor_value *val)
+{
+	int ret;
+
+	if (chan != SENSOR_CHAN_VOLTAGE) {
+		return -ENOTSUP;
+	}
+
+	switch ((uint16_t)attr) {
+	case SENSOR_ATTR_LOWER_THRESH:
+	case SENSOR_ATTR_UPPER_THRESH:
+	case SENSOR_ATTR_LOWER_VOLTAGE_THRESH:
+	case SENSOR_ATTR_UPPER_VOLTAGE_THRESH: {
+		/* Are we setting upper threshold? */
+		bool is_upper =
+			attr == SENSOR_ATTR_UPPER_THRESH ||
+			(uint16_t)attr == SENSOR_ATTR_UPPER_VOLTAGE_THRESH;
+		/* Are we setting threshold in millivolts? */
+		bool is_mv =
+			(uint16_t)attr == SENSOR_ATTR_LOWER_VOLTAGE_THRESH ||
+			(uint16_t)attr == SENSOR_ATTR_UPPER_VOLTAGE_THRESH;
+
+		ret = adc_cmp_npcx_set_threshold(dev, is_upper, is_mv,
+						val->val1);
+		break;
+		}
+	case SENSOR_ATTR_ALERT: {
+		const struct adc_cmp_npcx_config *const config = dev->config;
+
+		ret = adc_npcx_threshold_ctrl_enable(config->adc,
+						config->th_sel, !!val->val1);
+		break;
+		}
+	default:
+		ret = -ENOTSUP;
+	}
+	return ret;
+}
+
+static int adc_cmp_npcx_attr_get(const struct device *dev,
+			  enum sensor_channel chan,
+			  enum sensor_attribute attr,
+			  struct sensor_value *val)
+{
+	return -ENOTSUP;
+}
+
+static int adc_cmp_npcx_trigger_set(const struct device *dev,
+			     const struct sensor_trigger *trig,
+			     sensor_trigger_handler_t handler)
+{
+	const struct adc_cmp_npcx_config *const config = dev->config;
+	struct adc_cmp_npcx_data *data = dev->data;
+	struct adc_npcx_threshold_param param;
+
+	if (trig == NULL || handler == NULL) {
+		return -EINVAL;
+	}
+
+	if (trig->type != SENSOR_TRIG_THRESHOLD ||
+	    trig->chan != SENSOR_CHAN_VOLTAGE) {
+		return -ENOTSUP;
+	}
+
+	data->handler = handler;
+
+	param.type = ADC_NPCX_THRESHOLD_PARAM_WORK;
+	param.val = (uint32_t)&data->work;
+	return adc_npcx_threshold_ctrl_set_param(config->adc, config->th_sel,
+						&param);
+}
+
+static int adc_cmp_npcx_sample_fetch(const struct device *dev,
+			      enum sensor_channel chan)
+{
+	return -ENOTSUP;
+}
+
+static int adc_cmp_npcx_channel_get(const struct device *dev,
+			     enum sensor_channel chan,
+			     struct sensor_value *val)
+{
+	return -ENOTSUP;
+}
```

Comment:
Is it possible to allow the user to query if the voltage is above/below the trigger threshold?